### PR TITLE
virtio-mem: Cleanup the code based on specification

### LINKED ISSUE #1883: Verify Cloud Hypervisor conforms to virtio-mem specification
The VIRTIO spec has recently merged the specification related to `virtio-mem`, therefore it would be good to verify that our current code complies with the specification.
Reference to the `virtio-mem` specification PR.

### PR DESCRIPTION
**Pull request by sboeuf**
Relying on the virtio-mem specification, this PR performs a few simplifications. At the same time, it cleans up the code to make it more readable.

Fixes #1883

### CODE REVIEW

**Comment by sboeuf** [comment]
Last commit in this PR includes the intended fix from https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1909

File: virtio-devices/src/mem.rs [hunk 1]
```diff
@@ -119,6 +119,8 @@ pub enum Error {
     ResizeInval(String),
     // Fail to resize trigger
     ResizeTriggerFail(DeviceError),
+    // Invalid configuration
+    InvalidConfiguration(String),
```
Comment (rbradford) [inline virtio-devices/src/mem.rs:123]: Never put a String into and error like this. (I see the code already has that and it's bad.) Create a new `ValidateError` using thiserror/anyhow.

**Review by rbradford — COMMENTED** [review]
Thanks for the cleanup!

File: virtio-devices/src/mem.rs (hunk 1 above)
Comment (sboeuf) [reply virtio-devices/src/mem.rs:123]: Ok I will update :)

**Review by likebreath — APPROVED** [review]
Net contribution of removing 3 lines of code. Rob must be happy about the PR. ;) 

Good for me to learn some new about Rust and the code base. Thank you!

**Comment by sboeuf** [comment]
@rbradford I reorganized the commits, PTAL!

File: virtio-devices/src/mem.rs [hunk 2]
```diff
@@ -433,16 +435,21 @@ impl MemEpollHandler {
...
+            let res = unsafe {
+                libc::madvise(
+                    (self.host_addr + offset) as *mut libc::c_void,
+                    size as libc::size_t,
+                    libc::MADV_DONTNEED,
+                )
+            };
+            if res != 0 {
+                error!("madvise get error {}", io::Error::last_os_error());
+                return Err(Error::DiscardMemoryRange);
```
Comment (rbradford) [inline virtio-devices/src/mem.rs:451]: You lose the errno here.

File: virtio-devices/src/mem.rs [hunk 3]
```diff
@@ -433,16 +435,21 @@ impl MemEpollHandler {
...
+        if !self.hugepages {
+            let res = unsafe {
+                libc::madvise(
+                    (self.host_addr + offset) as *mut libc::c_void,
+                    size as libc::size_t,
+                    libc::MADV_DONTNEED,
+                )
+            };
+            if res != 0 {
+                error!("madvise get error {}", io::Error::last_os_error());
```
Comment (rbradford) [inline virtio-devices/src/mem.rs:450]: The English of this error message could be improved.

File: virtio-devices/src/mem.rs [hunk 4]
```diff
@@ -122,6 +122,8 @@ pub enum Error {
     ResizeTriggerFail(DeviceError),
     // Invalid configuration
     ValidateError(anyhow::Error),
+    // Failed discarding memory range
+    DiscardMemoryRange,
```
Comment (rbradford) [inline virtio-devices/src/mem.rs:126]: Include the std::io::Error.

File: virtio-devices/src/mem.rs [hunk 5]
```diff
@@ -357,197 +417,121 @@ struct MemEpollHandler {
...
+                libc::fallocate64(
+                    fd,
+                    libc::FALLOC_FL_PUNCH_HOLE | libc::FALLOC_FL_KEEP_SIZE,
+                    offset as libc::off64_t,
+                    size as libc::off64_t,
+                )
+            };
+            if res != 0 {
+                error!("fallocate64 get error {}", io::Error::last_os_error());
+                return Err(Error::DiscardMemoryRange);
```
Comment (rbradford) [inline virtio-devices/src/mem.rs:435]: Improve the error message here and propagate the errno.

File: virtio-devices/src/mem.rs [hunk 6]
```diff
@@ -357,197 +417,121 @@ struct MemEpollHandler {
...
+        let res = unsafe {
+            libc::madvise(
+                (self.host_addr + offset) as *mut libc::c_void,
+                size as libc::size_t,
+                libc::MADV_DONTNEED,
+            )
+        };
+        if res != 0 {
+            error!("madvise get error {}", io::Error::last_os_error());
+            return Err(Error::DiscardMemoryRange);
```
Comment (rbradford) [inline virtio-devices/src/mem.rs]: Ditto.

**Review by rbradford — APPROVED** [review]

**Comment by davidhildenbrand** [comment]
Regarding huge pages: you actually want fallocate(FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE) instead.

However, that won't be sufficient because anything file-based (IOW: not anonymous memory) requires more thought, especially:
- Memory consumption of MAP_PRIVATE mappings behave in a weird way due to COW. MAP_SHARED helps.
- Hugetlbfs reservations are nasty when you actually want sparse mappings. MAP_NORESERVE helps.
- With restricted resources (e.g., hugetlbfs) that cannot go to swap, we better do prepopulation (e.g., fallocate()) when plugging blocks and fail in a nice way (return -EBUSY to the VM) than crashing later via SIGBUS.
- There is no shared zeropage, so even any read will populate memory. At least Linux guests never read unplugged memory. One must be very careful in the hypervisor.

I'm currently in the process of solving the puzzle in QEMU, but it will for sure take a while until everything has been sorted out.
 Just FYI :)

**Comment by sboeuf** [comment]
@davidhildenbrand thanks for the heads up, please keep us posted on your findings and we'll update accordingly :)

**Pull request merged**